Add WEBGL_multi_draw extension

shrekshao opened the pull request:
Add WebGL extension WEBGL_multi_draw to emscripten.

Tested 4 functions locally with a test c src file modified from `webgl_draw_triangle.c`. (The test seems more like a webgl feature test so I didn't add to the test suite)

Having some extra questions:
* HEAPU8.subarray
  * Looking at https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html?highlight=subarray#access-memory-from-javascript My understand is that we need to create a new typed array few in case there's growth in memory.
  * But https://github.com/emscripten-core/emscripten/blob/master/src/library_webgl.js#L1812 use the HEAPU8.subarray directly. Wonder if creating a view is necessary or not here.
* Using impl of WEBGL_draw_instanced_base_vertex_base_instance as the reference, it seems like only WEBGL specific version of the draw call will be exported (in c src code it would probably be called around some `#ifdef EMSCRIPTEN` macro ). Doesn't seem like the native glDraw*BaseInstanceEXT etc would be exported. Is it the usual pattern for implementing WebGL extensions? (or is it hard to export native GL version functions etc.)
* About extension enabling
  * WEBGL_multi_draw is available for both WebGL 1 and 2. For chrome impl there's no overhead for enabling it, but not sure about other browsers. So didn't add it to automatically enabling extension list. Hope current change is valid.

shrekshao commented:
Wonder if the test_closure_full_js_library failures due to `ctx.multiDrawWebgl` access?

kripken commented on src/library_html5_webgl.js:
```diff
@@ -363,6 +363,9 @@ var LibraryHtml5WebGL = {
 
     var ext = context.GLctx.getExtension(extString);
+    if (extString == 'WEBGL_multi_draw') {
+      context.GLctx.multiDrawWebgl = ext;
```
why do we need this for this extension but not others? a comment could help here.

shrekshao replied:
comments added

kripken commented on src/library_webgl.js:
```diff
@@ -3594,6 +3594,59 @@ var LibraryGL = {
   },
 
+  glMultiDrawArraysWEBGL: function(mode, firsts, firstsOffset, counts, countsOffset, drawcount) {
+    var firstsList = new Int32Array(HEAPU8.subarray(firsts, firsts + drawcount * 4));
```
is the `new Int32Array()` call needed? That creates a copy, and I'm not sure why.

kenrussell replied:
Is there a way to access the heap as an Int32Array?

kripken replied:
Yes, `HEAP32` is a view with that type. So I think `HEAP32.subarray(first >> 2, first + drawcount * 4 >> 2)` would work?

shrekshao replied:
Looking at https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html?highlight=subarray#access-memory-from-javascript, my understanding is that if there's `-s ALLOW_MEMORY_GROWTH=1`, and the `firsts` pointer could become invalid?

kenrussell replied:
These pointers are just integer offsets into the WASM heap - they don't become invalid. 

I'm not 100% sure how memory growth works but the example in https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html?highlight=subarray#access-memory-from-javascript looks broken to me - the copy's being made out of the already-potentially-invalidated HEAP8.

I think you should ignore that document and just use the HEAP32 view as @kripken suggests above.

kripken replied:
Is the view retained by the called API? AFAIK old WebGL APIs don't do that, so growth is not a concern. But if this one does save the view, then it could be invalidated, and a copy is necessary.

kripken replied:
@kenrussell about the example, I think it's valid: malloc may invalidate `HEAPU8`, but we automatically recreate those global views when that happens. So the only risk is user views that we do not handle, that users must update themselves.

shrekshao replied:
Seems a copy would happen in the WebGL API impl. So creating a copy view here seems unnecessary to me
https://source.chromium.org/chromium/chromium/src/+/master:gpu/command_buffer/service/multi_draw_manager.cc;l=85?originalUrl=https:%2F%2Fcs.chromium.org%2F

kenrussell replied:
@kripken OK, thanks - the example's still a little confusing because it's the *next* call to func() which will invalidate the view returned from an earlier call.

@shrekshao agree, the copy is unnecessary.

shrekshao commented:
> Does the extension work in chrome beta (which is what we have on our CI)?

Yes, behind the --enable-webgl-draft-extensions flag

Let me refine and add the tests later

shrekshao commented:
Update a WEBGL_multi_draw specifc test (originally planned to add a kind of general test for webgl extensions but given after frequently running into small issues due to unfamiliar with the test suit)

One question is that the browser seems to be launched before entered certain test function in `test_browser.py`. How shall we add an extra command line arg for a specific test? (We can add to test for is_chrome if it wouldn't create any extra problem)

Also I realized that the extension might only be available when there's a physical gpu on the bot we are running on. The test simply uses regular gl.draw if the extension is not available (and will still pass the ref image test).

kenrussell commented on tests/webgl_multi_draw_test.c:
```diff
@@ -0,0 +1,148 @@
+    GLint firsts[] = {0};
+    GLsizei counts[] = {3};
+    GLsizei instanceCounts[] = {1};
+    GLint offsets[] = {0};
```
Could you modify this test to perform two draw calls via the MultiDraw APIs, perhaps drawing a rectangle instead of a single triangle? A bug could easily slip through where the implementation would execute only the first draw call in these lists.

shrekshao replied:
Fixed. It did help catch a bug 😓

kenrussell commented on tests/webgl_multi_draw_test.c:
```diff
@@ -0,0 +1,148 @@
+  
+  if (!extAvailable) {
+    glClear(GL_COLOR_BUFFER_BIT);
+    glDrawArrays(GL_TRIANGLES, 0, 3);
```
Is there a mechanism for skipping tests if they're not supported? This test would be more robust if it didn't produce a result if the extension was missing.

Is this an example of skipping a test in this harness?
https://github.com/emscripten-core/emscripten/blob/master/tests/canvas_animate_resize.cpp#L135

shrekshao replied:
Thanks for the pointer. Skip the test now when extension is not supported.

kenrussell commented on tests/test_browser.py:
```diff
@@ -4260,6 +4260,20 @@ def test_webgl_offscreen_canvas_only_in_pthread(self):
+  # We might want to append the --enable-webgl-draft-extensions to the EMTEST_BROWSER env arg.
+  @requires_graphics_hardware
+  def test_webgl_multi_draw(self):
+    self.btest('webgl_multi_draw_test.c', reference='webgl_array_of_structs_uniform.png',
```
While understanding you're trying to reuse an existing reference image here, I think a new image is warranted here per comments below.

kripken commented on src/library_html5_webgl.js:
```diff
@@ -363,6 +363,14 @@ var LibraryHtml5WebGL = {
     var ext = context.GLctx.getExtension(extString);
+    if (extString == 'WEBGL_multi_draw') {
+      // WEBGL_multi_draw is targeted at both WebGL 1 and WebGL 2.
+      // We now didn't list WEBGL_multi_draw as the automatically enabled extension now.
```
```suggestion
      // We currently don't list WEBGL_multi_draw as an automatically enabled extension,
```

kripken commented on src/library_html5_webgl.js:
```diff
@@ -363,6 +363,14 @@ var LibraryHtml5WebGL = {
+    if (extString == 'WEBGL_multi_draw') {
+      // WEBGL_multi_draw is targeted at both WebGL 1 and WebGL 2.
+      // We now didn't list WEBGL_multi_draw as the automatically enabled extension now.
+      // As there's only one browser implementation so we don't know if enabling it would
```
```suggestion
      // as there's only one browser implementation, so we don't know if enabling it would
```

kripken commented on src/library_html5_webgl.js:
```diff
@@ -363,6 +363,14 @@ var LibraryHtml5WebGL = {
+      // We now didn't list WEBGL_multi_draw as the automatically enabled extension now.
+      // As there's only one browser implementation so we don't know if enabling it would
+      // have extra overhead yet. This extension exposes extra functions so we need to attach
+      // them to multiDrawWebgl under GLctx object for future reference when calling.
```
is this aspect unique to this extension? If not then I'm surprised we didn't need something similar for other ones, and so there may be a better way than this. I can't think of one though. Anyhow, if @juj has a better idea later we can improve this then.

shrekshao replied:
> Is this aspect unique to this extension?
Yes. `WEBGL_multi_draw` and `WEBGL_*_base_vertex_base_instance` are still in draft. We add the implementation to hopefully get more clients feedback like unity and skia to help push forward these extensions to community approval.

Enabling this extension automatically when available (like most of the extensions) has no overhead in chromium. And that's for sure. We don't know for Firefox just because there's no implementation yet.

In https://github.com/emscripten-core/emscripten/pull/11054 @juj refactored some extensions that are only available to WebGL 1 context (or only available to WebGL 2 context) to have separate explicit enabling logic to prevent redundant code size. (say the user only targets at WebGL 2 context, having the code auto enabling extensions only available to WebGL 1 makes no sense)
But `WEBGL_multi_draw` is available to both WebGL 1 and 2 context so I don't think this falls into that category.

Maybe I'm making things overcomplicated here...

juj replied:
To follow the same pattern as the other extensions above (ANGLE_instanced_arrays, OES_vertex_array_object, etc.), this should not be handled here like this, as it would increase code size for code that does not use multi_draw.

Instead, let's create a new function `_webgl_enable_WEBGL_multi_draw`, and have

```js
    if (extString == 'WEBGL_multi_draw') __webgl_enable_WEBGL_multi_draw(GLctx);
```

and then also have a dedicated function
```js
  emscripten_webgl_enable_WEBGL_multi_draw__deps: ['_webgl_enable_WEBGL_multi_draw'],
  emscripten_webgl_enable_WEBGL_multi_draw: function(ctx) {
    return __webgl_enable_WEBGL_multi_draw(GL.contexts[ctx].GLctx);
  },
```
and in html5_webgl.h,
```h
extern EM_BOOL emscripten_webgl_enable_WEBGL_multi_draw(EMSCRIPTEN_WEBGL_CONTEXT_HANDLE context);
```

That way introducing the extension will not promote code size creep.

Also, even though enabling these extensions might cause slowdowns even when not used, we should add these to be automatically enabled in `GL.initExtensions()` function. That is, specifically, also call `__webgl_enable_WEBGL_multi_draw()` from within `GL.initExtensions()`. This is because that is the exact reason that `GL_SUPPORT_AUTOMATIC_ENABLE_EXTENSIONS` exists: it should enable all extensions to emulate desktop GLES behavior. (codebases that worry about such performance degradation should be building without that flag enabled)

juj replied:
> In #11054 @juj refactored some extensions that are only available to WebGL 1 context (or only available to WebGL 2 context) to have separate explicit enabling logic to prevent redundant code size. (say the user only targets at WebGL 2 context, having the code auto enabling extensions only available to WebGL 1 makes no sense)
But WEBGL_multi_draw is available to both WebGL 1 and 2 context so I don't think this falls into that category.

Specifically, the logic to avoid extra code size is not just about whether user is doing WebGL 1 only, or WebGL 2 only, but also whether user is doing WebGL2-without-multi-draw: i.e. if user wants to target WebGL 1 and/or 2, but does not care about the multi_draw feature at all, we should not be emitting extra code size to regress their build output sizes.

kenrussell replied:
@shrekshao could you please address this feedback next?

shrekshao replied:
Thanks for providing this detailed explanation. Sorry I missed these so was a bit confused. But thanks for highlighting them in the latest comment.
Fixed this. Also the one for base_vertex_base_instance seems to be missing. Fixed as well.
This commit is specially fixing this extension enabling issue: https://github.com/emscripten-core/emscripten/pull/11650/commits/27f3f8e12b2cbda50bbffbead01988b3ece1cc7b

kenrussell commented on tests/webgl_multi_draw_test.c:
```diff
@@ -0,0 +1,162 @@
+
+  glClearColor(0.3f,0.3f,0.3f,1);
+  
+  if (!extAvailable) {
```
It would be clearer if this block were right after the definition of the extAvailable variable above. Since the if-block returns 0, the else block isn't necessary, so all of that code doesn't need to be indented.

```
GLboolean extAvailable = ...;
if (!extAvailable) {
  // ...
  return 0;
}
```

kenrussell replied:
Would you consider again moving this block higher up, right after "GLboolean extAvailable = ..."?

shrekshao replied:
Sorry, missed that one. Fixed now.

juj commented on src/library_webgl.js:
```diff
@@ -3599,6 +3599,59 @@ var LibraryGL = {
+  glMultiDrawElementsInstancedWEBGL: function(mode, counts, countsOffset, type, offsets, offsetsOffset, instanceCounts, instanceCountsOffset, drawcount) {
+    var countsList = HEAP32.subarray(counts >> 2, (counts >> 2) + drawcount);
+    var offsetsList = HEAP32.subarray(offsets >> 2, (offsets >> 2) + drawcount);
+    var instanceCountsList = HEAP32.subarray(instanceCounts >> 2, (instanceCounts >> 2) + drawcount);
```
All these `HEAP32.subarray` calls make me propose that the `WEBGL_multi_draw` extension be changed to a function signature form that takes an array+offset+length triplet, to avoid these costly subarrays. It would be strongly preferred that hot paths in WebGL are able to operate without generating temporary garbage.

kenrussell replied:
Looking back at the extensions, I just realized that they've already been phrased this way, and that these bindings aren't taking advantage of them. See other comments in this review.

juj commented on tests/webgl2_simple_enable_extensions.c:
```diff
@@ -55,11 +55,15 @@ int main()
     assert(emscripten_webgl_enable_WEBGL_draw_instanced_base_vertex_base_instance(context));
 
+  if (hasext(exts, "WEBGL_multi_draw"))
+    assert(emscripten_webgl_enable_extension(context, "WEBGL_multi_draw"));
```
This should be inside `WEBGL_SIMPLE_ENABLE_EXTENSION`.

juj commented:
>     * Looking at https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html?highlight=subarray#access-memory-from-javascript My understand is that we need to create a new typed array few in case there's growth in memory.

We do not need to care anything special about memory growth here, the existing views will always be valid.

juj commented:
>     * But https://github.com/emscripten-core/emscripten/blob/master/src/library_webgl.js#L1812 use the HEAPU8.subarray directly. Wonder if creating a view is necessary or not here.

(note that codepath is for impaired performance WebGL 1 contexts, that one trashes garbage, unlike in WebGL 2)

>     * Doesn't seem like the native glDraw*BaseInstanceEXT etc would be exported. Is it the usual pattern for implementing WebGL extensions?

Whenever the WEBGL specific extension is identical to the functionality to a GLES/GL specific extension, we can (and should!) also export the functions to be statically linked available under the GLES/GL function names. That will ease native codebases to port, without practical downsides.

juj commented:
>     * For chrome impl there's no overhead for enabling it, but not sure about other browsers. So didn't add it to automatically enabling extension list. Hope current change is valid.

We should definitely add it to be automatically enabled, so that `-s GL_SUPPORT_AUTOMATIC_ENABLE_EXTENSIONS=1` will then do the right thing.

juj commented:
Great work on this overall!

kenrussell commented on src/library_webgl.js:
```diff
@@ -3599,6 +3601,59 @@ var LibraryGL = {
+      firstsOffset,
+      countsList,
+      countsOffset,
+      drawcount);
```
Looking back at the multi-draw extension:
https://www.khronos.org/registry/webgl/extensions/WEBGL_multi_draw/
and the ANGLE extension which it exposes:
https://chromium.googlesource.com/angle/angle/+/master/extensions/ANGLE_multi_draw.txt

I just realized that the C signatures in this pull request need to be updated - pairs like "firsts / firstsOffset" should be just a single pointer in the C signature. See comment below for an example. All of the entry points in this PR must be fixed.

This one should be written this way (please test this):

```
glMultiDrawArraysANGLE: function(mode, firsts, counts, drawcount) {
    GLctx.multiDrawWebgl['multiDrawArraysWEBGL'](
      mode,
      HEAP32,
      firsts >> 2,
      HEAP32,
      counts >> 2,
      drawcount);
```

("firsts" and "counts" are pointers into the WASM heap, which are just 32-bit numbers representing byte offsets. They need to be converted into offsets into an Int32Array.)

kenrussell commented on system/include/webgl/webgl1_ext.h:
```diff
@@ -228,6 +228,16 @@ WEBGL_APICALL void GL_APIENTRY emscripten_webgl_getSupportedAstcProfiles(GLsizei
 #define GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT 0x8C4F
 
+// 40. https://www.khronos.org/registry/webgl/extensions/WEBGL_multi_draw/
+WEBGL_APICALL void GL_APIENTRY emscripten_glMultiDrawArraysWEBGL(GLenum mode, const GLint* firstsList, GLuint firstsOffset, const GLsizei* countsList, GLuint countsOffset, GLsizei drawcount);
```
These signatures need to be revised. Please compare to the native ANGLE extension:
https://chromium.googlesource.com/angle/angle/+/master/extensions/ANGLE_multi_draw.txt

for the WebGL multi-draw extension:
https://www.khronos.org/registry/webgl/extensions/WEBGL_multi_draw/

All of the JavaScript argument pairs like "firstsList / firstsOffset" should just be a single pointer - "GLint* firsts" - in the C signature.

The underlying bindings then pass the incoming pointer from WASM - which is just an offset into the heap - down into the JavaScript entry point, without creating any temporary typed arrays.

Another question that needs to be answered: should these be named according to the underlying ANGLE extension? In this case, emscripten_glMultiDrawArraysANGLE? If C code were being written and linking against ANGLE then it would be calling glMultiDrawArraysANGLE, so I think probably these entry points should be named with the ANGLE suffix rather than the WEBGL suffix.

juj commented on src/library_webgl.js:
```diff
@@ -3601,6 +3601,49 @@ var LibraryGL = {
+      instanceCounts >> 2,
+      drawcount);
+  },
+
```
Let's name these functions according to the extension spec, that is the convention we have taken with other extensions as well. So `glMultiDrawElementsInstancedWEBGL` and so forth.

Then, to enable portability, add aliases
```js
glMultiDrawElementsInstancedWEBGL__sig: 'viiiiii',
glMultiDrawElementsInstancedANGLE: 'glMultiDrawElementsInstancedWEBGL'
```
That will cause the functions to be exposed under both names.

juj commented on system/include/webgl/webgl1_ext.h:
```diff
@@ -228,6 +228,17 @@ WEBGL_APICALL void GL_APIENTRY emscripten_webgl_getSupportedAstcProfiles(GLsizei
+WEBGL_APICALL void GL_APIENTRY glMultiDrawArraysANGLE(GLenum mode, const GLint* firsts, const GLsizei* counts, GLsizei drawcount);
+WEBGL_APICALL void GL_APIENTRY glMultiDrawArraysInstancedANGLE(GLenum mode, const GLint* firsts, const GLsizei* counts, const GLsizei* instanceCounts, GLsizei drawcount);
+WEBGL_APICALL void GL_APIENTRY glMultiDrawElementsANGLE(GLenum mode, const GLsizei* counts, GLenum type, const GLsizei* offsets, GLsizei drawcount);
+WEBGL_APICALL void GL_APIENTRY glMultiDrawElementsInstancedANGLE(GLenum mode, const GLsizei* counts, GLenum type, const GLsizei* offsets, const GLsizei* instanceCounts, GLsizei drawcount);
```
Here also let's name these with `WEBGL` suffix. I believe the `ANGLE` function signatures should be placed in `GLES2/gl2ext.h` header, as they are GLES extensions. I.e. people who compile GLES2/GLES3 codebases and target this extension without targeting WebGL headers would not be including `webgl1_ext.h` header file.

shrekshao commented:
Addressed review comments.
Didn't figure out what trigger the failure of `test_minimal_runtime_code_size` and `test_wasm_builtin_simd` (yet...). Would like to hear some thoughts...

juj commented on src/library_html5_webgl.js:
```diff
@@ -348,6 +349,8 @@ var LibraryHtml5WebGL = {
     if (extString == 'WEBGL_draw_instanced_base_vertex_base_instance') __webgl_enable_WEBGL_draw_instanced_base_vertex_base_instance(GLctx);
 #endif
 
+    if (extString == 'WEBGL_multi_draw') __webgl_enable_WEBGL_multi_draw(GLctx);
```
Oh, `WEBGL_multi_draw` should be present in the list below inside `ASSERTIONS`

juj commented on src/library_html5_webgl.js:
```diff
@@ -323,6 +323,7 @@ var LibraryHtml5WebGL = {
 #if MAX_WEBGL_VERSION >= 2
     '_webgl_enable_WEBGL_draw_instanced_base_vertex_base_instance',
 #endif
+    '_webgl_enable_WEBGL_multi_draw',
```
Can you gate all of the five above deps to be present only `#if GL_SUPPORT_SIMPLE_ENABLE_EXTENSIONS` - looks like that was missing from before. Not sure, but that might help with the minimal runtime code size test.

juj commented:
For the `test_minimal_runtime_code_size` failure, can you try adding `'-s', 'GL_SUPPORT_SIMPLE_ENABLE_EXTENSIONS=0',` around line 9285 in `tests/test_other.py`? (even if that will not fix it, that will be good to have there, looks like it is missing)

juj commented:
Another thing to try for the tests is to rebase against latest master, in case something was fixed up there in the meanwhile.

juj commented:
> Still not 100% sure about whether developers will expect to see WEBGL suffixes for these entry points instead of ANGLE versions, but let's get this out there and gather developer feedback.

Generally we carry both, with WEBGL suffixes containing the implementation, and GLES2/GLES3/Desktop GL extension names aliasing to the WEBGL implementation. WEBGL names in webgl_ext.h headers, and the GLES2/GLES3/Desktop GL extension names in their respective extension headers.

juj commented:
That is quality work, landed!

Squareys commented:
Just a quick hint that this extension cannot be used in combination with `-pthread`, as it expects non-shared array buffers as input.

As far as I can tell, there is no way to have both a block of shared and a block of unshared memory in WebAssembly, rahter it's always either or. A solution could be to copy the relevant parts of memory into a new buffer for calling the gl function.
Better ideas very appreciated!

kenrussell commented:
This was an oversight when specifying the WEBGL_multi_draw extension; it'll be revised to allow the incoming typed arrays to come from a SharedArrayBuffer. Thanks for raising this issue.

Squareys commented:
@kenrussell Oh amazing, so this might be eventually changed in Browsers rather than requiring a change emscripten, that's great!
